### PR TITLE
Added support for Alter Table `ADD COLUMNS` `DROP COLUMNS` `RENAME COLUMNS` `DROP CONSTRAINTS`

### PR DESCRIPTION
**Pull request by sundarshankar89**
Implement Alter Table

### CODE REVIEW

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/DataTypeGenerator.scala [hunk 1]
```diff
@@ -23,7 +23,7 @@ object DataTypeGenerator {
     case ir.DecimalType(precision, scale) =>
       val arguments = precision.toSeq ++ scale.toSeq
       if (arguments.isEmpty) {
-        "DECIMAL"
+        "DECIMAL(38, 0)"
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/generators/sql/DataTypeGenerator.scala:26]:
```suggestion
        "DECIMAL(10, 0)"
```

> p: Optional maximum precision (total number of digits) of the number between 1 and 38. The default is 10. s: Optional scale of the number between 0 and p. The number of digits to the right of the decimal point. The default is 0.



https://docs.databricks.com/en/sql/language-manual/data-types/decimal-type.html

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala [hunk 2]
```diff
@@ -34,12 +34,65 @@ class LogicalPlanGenerator(val expr: ExpressionGenerator, val explicitDistinct:
...
+    val operation = buildAlter(ctx, a.alterations)
+    s"ALTER TABLE ${a.tableName} $operation"
+  }
+
+  private def buildAlter(ctx: GeneratorContext, alteration: Seq[ir.TableAlteration]): String = {
+    alteration match {
+      case a: Seq[_] => buildTableAlteration(ctx, a)
+      case _ => ""
+    }
+  }
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala:54]: i don't understand it. what is the purpose of this?

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala [hunk 3]
```diff
@@ -34,12 +34,65 @@ class LogicalPlanGenerator(val expr: ExpressionGenerator, val explicitDistinct:
...
+
+  private def buildTableAlteration(ctx: GeneratorContext, c: Seq[_]): String = {
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala:56]: don't use `Seq[_]` here - we need concrete types in generator.

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala [hunk 4]
```diff
@@ -34,12 +34,65 @@ class LogicalPlanGenerator(val expr: ExpressionGenerator, val explicitDistinct:
...
+
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala:57]: remove blank lines

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala [hunk 5]
```diff
@@ -34,12 +34,65 @@ class LogicalPlanGenerator(val expr: ExpressionGenerator, val explicitDistinct:
...
+    c match {
+      case a: Seq[_] =>
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala:59]: `case a: Seq[_] =>` is bad code and it's not clear what it does.

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala [hunk 6]
```diff
@@ -34,12 +34,65 @@ class LogicalPlanGenerator(val expr: ExpressionGenerator, val explicitDistinct:
...
+        if (a.headOption.exists(_.isInstanceOf[ir.AddColumn])) {
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala:60]: rewrite this with 

```
something.map {
  case ir.AddColumn(cols) => s"ADD COLUMN ${cols.map(columnDeclaration(ctx, _).mkString(", ")}"
  case ir.DropColumns(names) => ...
  case ir.RenameColumn(old, new) => ...
}
```

File: core/src/test/scala/com/databricks/labs/remorph/transpilers/SnowflakeToDatabricksTranspilerTest.scala [hunk 7]
```diff
@@ -17,6 +17,48 @@ class SnowflakeToDatabricksTranspilerTest extends AnyWordSpec with TranspilerTes
 
   }
 
+  "Snowflake Alter commands" should {
+
+    "transpile Single ADD Columns" in {
+      "ALTER TABLE t1 ADD COLUMN c1 INTEGER;" transpilesTo (
```
Comment (nfx) [inline core/src/test/scala/com/databricks/labs/remorph/transpilers/SnowflakeToDatabricksTranspilerTest.scala:23]:
```suggestion
    "ALTER TABLE t1 ADD COLUMN c1 INTEGER" in {
      "ALTER TABLE t1 ADD COLUMN c1 INTEGER" transpilesTo (
```
use the same name for in block as the source query. we'll add more syntactic sugar later

**Review by nfx — CHANGES_REQUESTED** [review]

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/DataTypeGenerator.scala (hunk 1 above)
Comment (sundarshankar89) [reply core/src/main/scala/com/databricks/labs/remorph/generators/sql/DataTypeGenerator.scala:26]: Reverted, and changed in the Parser Side to accommodate Snow behavior to keep the generation consistent with our python parser.

https://docs.snowflake.com/en/sql-reference/data-types-numeric

File: core/src/test/scala/com/databricks/labs/remorph/transpilers/SnowflakeToDatabricksTranspilerTest.scala (hunk 7 above)
Comment (sundarshankar89) [reply core/src/test/scala/com/databricks/labs/remorph/transpilers/SnowflakeToDatabricksTranspilerTest.scala:23]: Done

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala (hunk 2 above)
Comment (sundarshankar89) [reply core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala]: Fixed it.

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala (hunk 4 above)
Comment (sundarshankar89) [reply core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala:59]: Fixed it

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala (hunk 5 above)
Comment (sundarshankar89) [reply core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala]: Fixed it

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala (hunk 4 above)
Comment (nfx) [reply core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala:59]: no, blank is still there.

File: core/src/main/scala/com/databricks/labs/remorph/parsers/snowflake/SnowflakeTypeBuilder.scala [hunk 8]
```diff
@@ -61,8 +61,10 @@ class SnowflakeTypeBuilder {
 
   private def decimal(c: DataTypeContext) = {
     val nums = c.num().asScala
-    val precision = nums.headOption.map(_.getText.toInt)
-    val scale = nums.drop(1).headOption.map(_.getText.toInt)
+    // https://docs.snowflake.com/en/sql-reference/data-types-numeric#number
+    // Per Docs defaulting the precision to 38 and scale to 0
+    val precision = nums.headOption.map(_.getText.toInt).getOrElse("38".toInt)
```
[the remark this answers is not in the document]
Comment (sundarshankar89) [reply core/src/main/scala/com/databricks/labs/remorph/parsers/snowflake/SnowflakeTypeBuilder.scala:66]: Brain Fade!!! Fixed it.

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala [hunk 9]
```diff
@@ -34,12 +34,71 @@ class LogicalPlanGenerator(val expr: ExpressionGenerator, val explicitDistinct:
...
     case x => throw unknown(x)
   }
 
+  private def alterTable(ctx: GeneratorContext, a: ir.AlterTableCommand): String = {
+    val operation = buildTableAlteration(ctx, a.alterations)
+    s"ALTER TABLE ${a.tableName} $operation"
+  }
+
+  private def buildTableAlteration(ctx: GeneratorContext, alterations: Seq[ir.TableAlteration]): String = {
+    alterations.headOption match {
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala:50]: it's not clear why do you need a headOption. alterations is a seq, so you have to do:

```
alterations.map {
   case ir.AddColumn(decl) => columnDeclaration(decl)
   case ir.AddConstraint(columnName, constraint) => ...
   ..
}
```

File: core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala [hunk 10]
```diff
@@ -34,12 +34,71 @@ class LogicalPlanGenerator(val expr: ExpressionGenerator, val explicitDistinct:
...
+      case Some(a: ir.AddColumn) => buildAddColumn(ctx, alterations.collect { case c: ir.AddColumn => c })
```
Comment (nfx) [inline core/src/main/scala/com/databricks/labs/remorph/generators/sql/LogicalPlanGenerator.scala:51]: this code is incorrect and is going to break when there's more than one alteration.

**Review by nfx — CHANGES_REQUESTED** [review]
this code is incorrect and is going to break when there's more than one alteration.

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**